File: src/test/java/hex/paymentstore/adapter/PaymentRepositoryCsvFileTest.java
package hex.paymentstore.adapter;

import hex.paymentstore.domain.BankAccountNumber;
import hex.paymentstore.domain.Payment;
import hex.paymentstore.domain.UserId;
import org.junit.jupiter.api.Test;

import java.io.File;
import java.io.IOException;
import java.math.BigDecimal;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;

class PaymentRepositoryCsvFileTest {
	
	@Test
	void should_add_payment() throws IOException {
		//given
		Payment payment = new Payment("id", BigDecimal.TEN, "PLN", new UserId("userId"), new BankAccountNumber("123"));
		
		// when
		Payment added = sut().add(payment);
		
		// then
		assertEquals(added, payment);
	}
	
	@Test
	void should_get_all_payments() throws IOException {
		//given
		Payment payment1 = new Payment("1", BigDecimal.TEN, "PLN", new UserId("userId"), new BankAccountNumber("123"));
		Payment payment2 = new Payment("2", BigDecimal.TEN, "PLN", new UserId("userId"), new BankAccountNumber("123"));
		
		PaymentRepositoryCsvFile sut = sut();
		sut.add(payment1);
		sut.add(payment2);
		
		// when
		List<Payment> payments = sut.getAll();
		// then
		assertTrue(payments.containsAll(List.of(payment1, payment2)));
	}
	
	@Test
	void should_find_payment() throws IOException {
		//given
		PaymentRepositoryCsvFile sut = sut();
		Payment added = sut.add(new Payment("id", BigDecimal.TEN, "PLN", new UserId("userId"), new BankAccountNumber("123")));
		
		// when
		Optional<Payment> found = sut.find(added.id());
		// then
		assertTrue(found.isPresent());
		assertEquals(found.get(), added);
	}
	
	@Test
	void should_not_find_payment() throws IOException {
		// when
		Optional<Payment> found = sut().find("notExistingPayment");
		// then
		assertTrue(found.isEmpty());
	}
	
	@Test
	void should_remove_payment() throws IOException {
		//given
		PaymentRepositoryCsvFile sut = sut();
		Payment added = sut.add(new Payment("id", BigDecimal.TEN, "PLN", new UserId("userId"), new BankAccountNumber("123")));
		
		// when
		sut.remove(added.id());
		// then
		assertTrue(sut.find(added.id()).isEmpty());
	}
	
	private PaymentRepositoryCsvFile sut() throws IOException {
		File tempFile = File.createTempFile("tmp", ".csv");
		tempFile.deleteOnExit();
		return new PaymentRepositoryCsvFile(tempFile.getPath());
	}
}

File: src/test/kotlin/hex/paymentstore/adapter/PaymentRepositoryCsvFileTest.kt
package hex.paymentstore.adapter

import hex.paymentstore.domain.BankAccountNumber
import hex.paymentstore.domain.Payment
import hex.paymentstore.domain.UserId
import io.kotest.core.spec.style.ShouldSpec
import io.kotest.matchers.collections.containAll
import io.kotest.matchers.should
import io.kotest.matchers.shouldBe
import java.math.BigDecimal

class PaymentRepositoryCsvFileTest : ShouldSpec({
	
	fun sut(): PaymentRepositoryCsvFile {
		val tempFile = createTempFile()
		tempFile.deleteOnExit()
		return PaymentRepositoryCsvFile(tempFile.path)
	}
	
	should("add payment") {
		val payment = Payment("id", BigDecimal.TEN, "PLN", UserId("userId"), BankAccountNumber("123"))
		
		// when
		val added = sut().add(payment)
		
		// then
		added shouldBe payment
	}
	
	should("get all payments") {
		val payment1 = Payment("1", BigDecimal.ONE, "PLN", UserId("userId"), BankAccountNumber("123"))
		val payment2 = Payment("2", BigDecimal.TEN, "PLN", UserId("userId"), BankAccountNumber("456"))
		val sut = sut()
		sut.add(payment1)
		sut.add(payment2)
		
		// when
		val payments = sut.getAll()
		// then
		payments should containAll(payment1, payment2)
	}
	
	should("get payment") {
		val payment = Payment("id", BigDecimal.TEN, "PLN", UserId("userId"), BankAccountNumber("123"))
		val sut = sut()
		sut.add(payment)
		
		// when
		val found = sut.find(payment.id!!)
		// then
		found shouldBe payment
	}
	
	should("not find payment") {
		// when
		val found = sut().find("notExistingPayment")
		// then
		found shouldBe null
	}
	
	should("remove payment") {
		val payment = Payment("id", BigDecimal.TEN, "PLN", UserId("userId"), BankAccountNumber("123"))
		val sut = sut()
		sut.add(payment)
		
		// when
		sut.remove(payment.id!!)
		// then
		sut.find(payment.id!!) shouldBe null
	}
})